### PR TITLE
Merge kp/support min max value float datetime into master for adding 

### LINKED ISSUE #415: Support minValue maxValue in the sdc library for Decimal, Date and DateTime values
This issue is a followup issue for #164

### PR DESCRIPTION
**Pull request by kunjan8794**
Closes #415

**Description**
Added support for decimal, date & dateTime data types for MinValue & MaxValue extensions

**Type**
Feature

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by kunjan8794** [comment]
@jingtang10 Just wanted your inputs regarding the unit tests. How should I do it here for different data types? I have added couple of them for the decimal data types. If you think thats apt, I will go ahead and write them for date and dateTime as well.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
you can define a comparison method on  the type `Type`:

```
operator fun Type.compareTo(value: Type?): Int {
  ...
}
```

and then in your validators you can just do this:

```
{ extension: Extension, answer: QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent
      ->
      extension.value > answer.value  // or extension.value < answer.value
    }
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.hl7.fhir.r4.model.Type
+
+operator fun Type.compareTo(value: Type?): Int {
+  if (value != null) {
+    when {
+      this.fhirType().equals("integer") && value.fhirType().equals("integer") -> {
+        if (value.primitiveValue().toInt() > this.primitiveValue().toInt()) return 1
+        if (value.primitiveValue().toInt() < this.primitiveValue().toInt()) return -1
+        if (value.primitiveValue().toInt() == this.primitiveValue().toInt()) return 0
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt:27]:
```
value.primitiveValue().toInt().compareTo(primitiveValue().toInt())
```

Same with the other types

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/*
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt:1]: please add tests for this class

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      }
+      this.fhirType().equals("dateTime") && value.fhirType().equals("dateTime") -> {
+        return this.dateTimeValue().value.compareTo(value.dateTimeValue().value)
+      }
+    }
+  }
+  return 0
+}
+
+private fun clearTimeFromDateValue(dateValue: Date): Date {
```
Comment (kunjan8794) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt:44]: @jingtang10 I have added this utility funciton here, but I am not sure if thats the correct place for this function. Can you please share your insights on this?

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt [hunk 4]
```diff
@@ -55,6 +56,29 @@ class MinValueConstraintValidatorTest {
     assertThat(validationResult.message).isEqualTo("Minimum value allowed is:10")
   }
 
+  @Test
+  fun shouldReturnInvalidResultForDecimal() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/MinValueConstraintValidatorTest.kt:60]:
```suggestion
  fun shouldReturnInvalidResult() {
```

so in terms of responsibilities, the `MinValueConstraintValidator` only checkes if something is less than the min value. It doesn't really care what type the value is. So in the tests, what we are testing is that it invokes the comparison correctly.

On the other hand, the `MoreTypeTest` needs to verify the comparison works with different types of value.

So here I think it makes more sense to remove the `ForDecimal` in the name of the test case.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by jingtang10** [comment]
Hey @kunjan8794 any update here? Let me know if anything isn't clear in my review. Thanks!

**Comment by jingtang10** [comment]
let's try to close this off soon since the pr was opend a while ago. :)

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  val calendarValue = Calendar.getInstance()
```
[the remark this answers is not in the document]
Comment (kunjan8794) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt:45]: @jingtang10 the Date(year, month, day) function has been deprecated. Should we still use it?

[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/validation/MoreTypes.kt:48]: I see - in that case it's fine.

**Review by jingtang10 — APPROVED** [review]
Thanks @kunjan8794 - just 3 final comments.

**Comment by jingtang10** [comment]
you may need to run `./gradlew spotlessApply`.

**Pull request merged**